Support for attention request

felipefs opened the pull request:
This pull contains the implementation of attention request for X11.

The added function was ```glfwRequestWindowAttention```.

The example ```simple.c``` was updated to demonstrate the functionality.

This patch is related with the issue https://github.com/glfw/glfw/issues/732.

elmindreda commented on src/cocoa_window.m:
```diff
@@ -1272,6 +1272,11 @@ void _glfwPlatformShowWindow(_GLFWwindow* window)
 
+void _glfwPlatformRequestWindowAttention(_GLFWwindow* window)
+{
+    [window->ns.object requestUserAttention:NSInformationalRequest];
```
`requestUserAttention:` is part of NSApplication, not NSWindow.

elmindreda commented:
Because some platforms (read macOS) have application-level attention requests, perhaps the more neutral `glfwRequestAttention` would be a better name?

felipefs commented:
I was wondering how we could approach that without breaking compatibility between platforms, since on Linux/Window, it will look like ```glfwRequestAttention(GLFWwindow *)``` and on macOS ```glfwRequestAttention(void)```.

elmindreda commented:
Have it take a window and document that it's a hint and not guaranteed to point out the specific window.

felipefs commented:
I've updated the function to do the following:
```objective-c
void _glfwPlatformRequestWindowAttention(_GLFWwindow* window)
{
    [NSApp requestUserAttention:NSInformationalRequest];
}
```
Also, I updated the documentation, adding a remark that on OSX the attention refers to the application, not the window.

Let me know if something is incorrect and I will fix 👍

elmindreda commented:
This looks almost ready to merge. Please don't add any test code or additional features to the `simple` example, though. It's mean to be just that and is [used in the documentation](http://www.glfw.org/docs/latest/quick_guide.html#quick_example).

felipefs commented:
Do you want me to remove the added code from ```simple.c```?

elmindreda commented:
Yes, please do. Instead you could for example request attention whenever the `joysticks` test detects a connection or disconnection.

elmindreda commented on src/x11_window.c:
```diff
@@ -2078,6 +2078,21 @@ void _glfwPlatformHideWindow(_GLFWwindow* window)
+    xev.xclient.data.l[0] = _NET_WM_STATE_ADD;
+    xev.xclient.data.l[1] = _glfw.x11.NET_WM_STATE_DEMANDS_ATTENTION;
+
+    XSendEvent(_glfw.x11.display, DefaultRootWindow(_glfw.x11.display), False, SubstructureRedirectMask | SubstructureNotifyMask, &xev);
```
This function could be made significantly shorter by calling the helper `sendEventToWM`.

elmindreda commented:
This has been merged as 412eb6a6113f71f42ad56d8dd982003ba7094870, with additional cleanup in baa9cd8968e1571cd10a7265b589d997febf611e and 2f5e2303380d084ab9b097765d34f98504324d65.
